test/hotspot/jtreg/compiler/jeandle/TestMakeNotEntrant.java

```diff
@@ -0,0 +1,148 @@
+    public final static int threadCount = 3;
+    public final static int makeNotEntrantCount = 3;
+    public static volatile boolean finish = false;
+
+    public static void main(String[] args) throws Exception {
+        if (args.length == 1 && args[0].equals("true")) {
+            List<String> jvmArgs = ManagementFactory.getRuntimeMXBean().getInputArguments();
+            List<String> cmdLine = new ArrayList<>();
+            cmdLine.addAll(jvmArgs);
+            cmdLine.add("TestMakeNotEntrant");
+            ProcessBuilder pb = ProcessTools.createLimitedTestJavaProcessBuilder(cmdLine);
+            OutputAnalyzer output = new OutputAnalyzer(pb.start());
+            output.shouldHaveExitValue(0);
+
+            String message = "TestMakeNotEntrant::add (4 bytes)   made not entrant";
+            checkOccurrence(output.getOutput(), message, makeNotEntrantCount);
+            return;
+        }
+
+        test();
+    }
+
+    public static void test() throws Exception {
+        for (int i = 0; i <= threadCount; i++) {
+            runThread();
+        }
+
+        Method m = getMethod("add", int.class, int.class);
+        int count = makeNotEntrantCount;
+        while(count > 0) {
+            if (wb.isMethodCompiled(m)) {
+                makeNotEntrant(m);
+                count--;
+            }
+        }
+        Thread.sleep(100);
+        finish = true;
+    }
+
+    public static Method getMethod(String methodName, Class<?>... parameterTypes) throws Exception {
+        return TestMakeNotEntrant.class.getDeclaredMethod(methodName, parameterTypes);
+    }
+
+    public static void makeNotEntrant(Method m) {
+        MetaAccessProvider metaAccess = JVMCI.getRuntime().getHostJVMCIBackend().getMetaAccess();
+        metaAccess.lookupJavaMethod(m).reprofile();
+    }
+
+    public static void checkOccurrence(String output, String str, int time) {
+        Pattern pattern = Pattern.compile(Pattern.quote(str));
+        Matcher matcher = pattern.matcher(output);
+        int count = 0;
+        while (matcher.find()) {
+            count++;
+        }
+        Asserts.assertEquals(count, time,
+                "The string \"" + str + "\" should appear " + time + " times, but appears " + count + " times.");
+    }
+
+    public static void runThread() {
+        Thread loopThread = new Thread(() -> {
+            loopTest();
+        });
+        loopThread.setDaemon(true);
+        loopThread.start();
+    }
+
+    public static void loopTest() {
+        while (!finish) {
+            /* patching verified_entry without reserved 5-byte header in x86_64, the following cases will occur.
+             * 1、SIGILL: exit_value not zero, can be observed.
+             * 2、SIGSEGV: exit_value not zero, can be observed.
+             * 3、Nothing happened but hurt header code. In "TestMakeNotEntrant::add" method, it will lose the add operation.
+             *    Thus, modify the return value, and then check it.
```

Comment:
```suggestion
            /* patching verified_entry without reserved 5-byte header in x86_64, the following cases will occur.
             * 1、SIGILL: The exit value of current process is not zero, can be observed.
             * 2、SIGSEGV: The exit value of current process is not zero, can be observed.
             * 3、Nothing happened but hurt header code. "TestMakeNotEntrant::add" method will lose the add operation.
             *    Thus, change the return value, and then check it.
```
I think this is a more clear version. But I still don't understand why we should change the return value. If patching verified_entry without reserved 5-byte, what will happen to the add instruction? Please explain it more.

Reply:
The compiled code of Method "TestMakeNotEntrant::add" is
```
  0x0000:   push   %rbp
  0x0001:   lea    (%rsi,%rdx,1),%eax
  0x0004:   pop    %rbp
  0x0005:   ret
```
so, the instruction `lea    (%rsi,%rdx,1)` will definitely be modified when patching verified_entry, and we can check the `eax` register to determine whether this instruction has been corrupted. However, a single check is not sufficient because no other instructions in this loop will modify the `eax` register. Therefore, after the first check, we change the return value and check it again.